documentation/jetty-documentation/src/main/asciidoc/programming-guide/arch-threads.adoc

```diff
@@ -0,0 +1,200 @@
+
+This mode solves the head-of-line blocking discussed in the xref:pg-arch-threads-execution-strategy-pc[`Produce-Consume` section], but suffers of other issue:
+
+* It is not CPU core cache efficient, as the data available to the producer thread will need to be accessed by another thread that likely won't have that data in its caches.
+* If the tasks take time to be run, the `Executor` queue may grow indefinitely.
+* A small latency is added to every task: the time it waits in the `Executor` queue.
+
+[[pg-arch-threads-execution-strategy-epc]]
+===== Execute-Produce-Consume
+In the `Execute-Produce-Consume` mode, the producer thread `Thread 1` loops to produce a task, then submits one internal task to an `Executor` to take over production on thread `Thread 2`, and then runs the task in `Thread 1`, and so on.
+
+[plantuml]
+----
+skinparam backgroundColor transparent
+
+compact concise "Thread 1" as WT1
+compact concise "Thread 2" as WT2
+compact concise "Thread 3" as WT3
+compact concise "Thread 4" as WT4
+hide time-axis
+
+@WT1
+0 is T1 #lightgreen
+1 is "Run T1" #dodgerblue
+5 is {hidden}
+
+@WT2
+1 is T2 #lightgreen
+2 is "Run T2" #dodgerblue
+4 is T5 #lightgreen
+5 is "Run T5" #dodgerblue
+10 is {hidden}
+
+@WT3
+2 is T3 #lightgreen
+3 is "Run T3" #dodgerblue
+6 is {hidden}
+
+@WT4
+3 is T4 #lightgreen
+4 is "Run T4" #dodgerblue
+8 is {hidden}
+
+----
+
+This mode may operate like xref:pg-arch-threads-execution-strategy-pc[`Produce-Consume`] when the take over production task run, for example, by thread `Thread 3` takes time to be executed (for example, in a busy server): thread `Thread 2` will produce one task and run it, then produce another task and run it, etc.
+By the time thread `Thread 3` takes over task production from `Thread 2`, all the work might already be done.
+
+This mode may also operate similarly to xref:pg-arch-threads-execution-strategy-pec[`Produce-Execute-Consume`] when the take over production task always finds a free CPU core immediately (for example, in a mostly idle server): thread `Thread 1` will produce a task, yield production to `Thread 2` while `Thread 1` is running the task; `Thread 2` will produce a task, yield production to `Thread 3` while `Thread 2` is running the task, etc.
+
+Differently from `Produce-Execute-Consume`, here production happens on different threads, but the advantage is that the task is run by the same thread that produced it (which is CPU core cache efficient).
+
+[[pg-arch-threads-execution-strategy-adaptive]]
+===== Adaptive Execution Strategy
+The modes of task consumption discussed above are captured by the `org.eclipse.jetty.util.thread.ExecutionStrategy` interface, with an additional implementation that also takes into account the behavior of the task when the task implements `Invocable`.
+
+For example, a task that declares itself as non-blocking could be consumed using the `Produce-Consume` mode, since there is no risk to stop production because the task will not block.
```

Comment:
```suggestion
For example, a task that declares itself as non-blocking can be consumed using the `Produce-Consume` mode, since there is no risk to stop production because the task will not block.
```

Reply:
Done.